### ISSUE TITLE
Extremely slow building of sandbox from image: dd with bs=31, use iflag=skip_bytes instead.

### ISSUE THREAD

**Issue by drhpc** [apptainer/singularity#4550]
I installed singularity 3.3.0-1 on a HPC cluster from source (I'm the admin) and am trying to convert a .sif to a sandbox to check how runtime performance can improve with --userns, avoiding the temporary extraction to $TMPDIR.

So I do this:
```
singularity pull shub://hariszaf/pema:v1
singularity build --sandbox pema_v1 pema_v1.sif
```
And then I wait … and wait … and notice that I am waiting for _this_ to finish:
```
dd bs=31 skip=1 if=pema_v1.sif of=/work/tmp/node002.2019-10-02-133002.rrztest.16897/sbuild-838622514/trim.squashfs667647721
```
This is excruciatingly slow. A copy in 31-byte blocks? This takes ages to copy 2G around on a parallel filesystem that easily gives you around 1G/s on a single stream.

Compare this:
```
-bash-4.2$ dd bs=1M skip=31 iflag=skip_bytes  if=pema_v1.sif  of=pema_v1.squashfs
2699+1 records in
2699+1 records out
2830450688 bytes (2.8 GB) copied, 3.69102 s, 767 MB/s
```
While still waiting for `singularity build` to finish, I created the raw squashfs
this way and ran it through `unsquashfs`. This is _a_lot_ faster. It's slower without bs=1M, but still orders of magnitude faster than working in 31-byte-blocks!

Please consider changing the inefficient `dd` call. I might end up having the users use the .sif image via
```
TMPDIR=/dev/shm singularity run --userns pema_v1.sif
```
anyway, as that only has 10 seconds overhead (obviously uses something more efficient than the build command), but maybe I want to offer them the sandbox directories, if user permissions don't cause trouble in there. And in an HPC context this wildly inefficient I/O just looks weird.

Is it maybe a bug that `singularity build` uses a different routine to extract the .sif than `singularity run --userns`? An interesting difference is that the latter breaks on SIGSTOP (interrupted system call on putting it into the backgroud in the shell), while the former is fine with that. Both could learn from each other, I suppose.

PS: While composing this report and testing things, the initial `build` advanced to 2.2G of extracted squashfs. It will finish _eventually_. In the end, it's silly having to make a copy of the whole image just to skip a header. If `unsquashfs` supported an input offset … and how did we arrive at 31 bytes to begin with? Not even 32, let alone a sane number like 256, 4K or so.

**Comment by dtrudg** [comment]
Unfortunately, `skip_bytes` is only available in coreutils 8.16 and above, from looking at https://github.com/coreutils/coreutils/commit/140eca1

We have to support RHEL6, which is using coreutils 8.4 - so simply inserting the `skip_bytes` flag is not possible here - more thought is needed on this.

@cclerget or @ikaneshiro may be able to weigh in on the userns run vs build difference

**Comment by dtrudg** [comment]
Also when you say:

>  trying to convert a .sif to a sandbox to check how runtime performance can improve with --userns, avoiding the temporary extraction to $TMPDIR

... are you expecting running from a sandbox to be faster than from a SIF? The SIF squashfs is loopback mounted, so there is no temporary extraction on a `singularity run` of a SIF. `$TMPDIR` is used during build to assemble the container rootfs from docker layers or other sources before it is converted into a SIF. Generally, on network parallel filesystems especially, running from a SIF is far faster than running from a sandbox as metadata operations are against a local loopback mount of squashfs rather than going out to the parallel fs for each file accessed in a sandbox dir.

**Comment by cclerget** [comment]
@drhpc @dctrud Run command use the dedicated extractor https://github.com/sylabs/singularity/blob/master/cmd/internal/cli/actions_linux.go#L47 without relying on dd which explains the difference

**Comment by ikaneshiro** [comment]
@drhpc @dctrud @cclerget This should have been fixed with https://github.com/sylabs/singularity/pull/3227 as that was intended to move extraction to using the dedicated extractor, but I neglected to cover the case for 2.x images

**Comment by drhpc** [comment]
@dctrud 
> Also when you say:
> 
> > trying to convert a .sif to a sandbox to check how runtime performance can improve  with --userns, avoiding the temporary extraction to $TMPDIR
> 
> ... are you expecting running from a sandbox to be faster than from a SIF? The SIF squashfs is loopback mounted

Except, it is not;-) With `--userns`, and especially in our setup without any `suid`, loop-mounts are not possible. The help is rather explicit about this:
```
  -u, --userns                 run container in a new user namespace,
                               allowing Singularity to run completely
                               unprivileged on recent kernels. This
                               disables some features of Singularity, for
                               example it only works with sandbox images.
```
It fails to mention that SIF files still work, as they are transparently converted to sandboxes (using TMPDIR).

**Comment by dtrudg** [comment]
@drhpc - ahh, from the original wording I misread that you were specifically using `--userns` to force the sandbox creation, not that you were trying to speed up things for userns.

Yes, with user namespace the loop mount is not possible, and the message is inaccurate here.

**Comment by drhpc** [comment]
Coreutils archaeology, just for fun:

@dctrud 
> Unfortunately, `skip_bytes` is only available in coreutils 8.16 and above, from looking at [coreutils/coreutils@140eca1](https://github.com/coreutils/coreutils/commit/140eca1)

This may be moot once the real fix uses the direct extractor, but for completeness, I'd suggest
```
dd bs=1 skip=31 count=97  if=pema_v1.sif of=pema_v1.squashfs.again
dd bs=128 skip=1 count=31 conv=notrunc oflag=append if=pema_v1.sif of=pema_v1.squashfs.again
dd bs=4K skip=1 count=255 conv=notrunc oflag=append if=pema_v1.sif of=pema_v1.squashfs.again
dd bs=1M skip=1 conv=notrunc oflag=append if=pema_v1.sif of=pema_v1.squashfs.again
```
(with some checking for actual image size if smaller than 1M).

I first stumbled over apparently obsolete syntax from fileutils:
```
[4.0.33]
* dd now accepts skip=nB and seek=nB, to advance past some number of bytes, n,
that need not be a multiple of the block size.
```
That gives me an invalid number. But there is append mode since 2005.

And of course, one could just write a simple proper program to do this, like the faster extraction method you already have.

**Comment by ikaneshiro** [comment]
Hey @drhpc, would be great if you could test out #4556 if you have the chance. This uses the faster extraction method and should fix your slow build issue.

**Comment by drhpc** [comment]
@ikaneshiro Hm, I tried getting a build from master to run, but mconfig now insists on cryptsetup. This is not installed on our compute nodes where I build things for the cluster and I don't see a reason why we would add it.

Could this be made optional, like it was for version 3.3? We run a squashfs image from RAM on the diskless nodes and install additional software (like singularity) into a prefix mounted over NFS. Of course I could also install cryptsetup that way, but what purpose would it serve with a non-suid install of singularity?

**Comment by drhpc** [comment]
I managed to avoid the issue by commenting out the
```
exit 2
```
in the config script. I tested it now and the extraction seems to be a lot more reasonable. It still takes its time on a network fs, but does not block with that crawling `dd` call.

Should I open an issue about allowing the build without cryptsetup? I see checks at runtime that handle incompatible cryptsetup and disable the functionality. It only seems logical to also treat the case of missing cryptsetup gracefully.
